packages/gatsby/cache-dir/ssr-develop-static-entry.js

```diff
@@ -10,6 +8,16 @@ import syncRequires from "$virtual/ssr-sync-requires"
 import { RouteAnnouncerProps } from "./route-announcer-props"
 import { ServerLocation, Router, isRedirect } from "@reach/router"
 
+let React
+let ReactDom
+if (process.env.CI) {
+  React = require(`react/cjs/react.production.min.js`)
+  ReactDom = require(`react-dom/cjs/react-dom-server.node.production.min.js`)
+} else {
+  React = require(`react`)
+  ReactDom = require(`react-dom/server`)
+}
+
```

Comment:
Would do you think about adjusting webpack config for it instead of having conditional imports (and mixing cjs with es imports)?

```js
    config.externals = [
      function (context, request, callback) {
        if (
          stage === `develop-html` &&
          isCI() &&
          process.env.GATSBY_EXPERIMENTAL_DEV_SSR
        ) {
          if (request === `react`) {
            callback(null, `commonjs react/cjs/react.production.min.js`)
            return
          } else if (request === `react-dom/server`) {
            callback(
              null,
              `react-dom/cjs/react-dom-server.node.production.min.js`
            )
            return
          }
        }

        const external = isExternal(request)
        if (external !== null) {
          callback(null, external)
        } else {
          callback()
        }
      },
    ]
```

( in https://github.com/gatsbyjs/gatsby/blob/8ff6245f44882d891bc6f291363019047cbdb55f/packages/gatsby/src/utils/webpack.config.js#L676-L686 )

Reply:
I'm 100% fine with conditional requires as they get removed during compilation and I'd assume having it in the rendering code is more discoverable than if it's in the webpack config.

Reply:
The other problem here _might_ be that's not the only place where we import `react-dom/server` - i.e. https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-plugin-glamor/src/gatsby-ssr.js and any plugin that use `replaceRenderer` will not use those swaps - if it is just not using prod builds - that's no big deal, worse if it cause weird conflicts like "multiple react versions" or something (and that might be more toward `react` than `react-dom/server`?)

Reply:
(and `glamor` plugin is just example - if there is one, there can be many :P)

Reply:
Oh, actually - wouldn't it not use prod `react` for user defined components now? Not sure how much perf there is in `react` core and how much of it is in `react-dom/server` - but this might actually be more performant to use webpack aliases/externals swaps :P

Reply:
after adding some logs in `node_modules/react` - we use both `react.prod` and `react.dev` with this PR as-is:

```
loading root react
Trace: 
    at Object.<anonymous> (/Users/misiek/dev/gatsby-starter-blog/node_modules/react/index.js:5:9)
    at Module._compile (internal/modules/cjs/loader.js:1138:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1158:10)
    at Module.load (internal/modules/cjs/loader.js:986:32)
    at Function.Module._load (internal/modules/cjs/loader.js:879:14)
    at Module.require (internal/modules/cjs/loader.js:1026:19)
    at require (internal/modules/cjs/helpers.js:72:18)
    at Object.<anonymous> (/Users/misiek/dev/gatsby-starter-blog/node_modules/@reach/router/index.js:8:14)
    at Module._compile (internal/modules/cjs/loader.js:1138:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1158:10)
loading react.development
Trace: 
    at Object.<anonymous> (/Users/misiek/dev/gatsby-starter-blog/node_modules/react/cjs/react.development.js:13:9)
    at Module._compile (internal/modules/cjs/loader.js:1138:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1158:10)
    at Module.load (internal/modules/cjs/loader.js:986:32)
    at Function.Module._load (internal/modules/cjs/loader.js:879:14)
    at Module.require (internal/modules/cjs/loader.js:1026:19)
    at require (internal/modules/cjs/helpers.js:72:18)
    at Object.<anonymous> (/Users/misiek/dev/gatsby-starter-blog/node_modules/react/index.js:10:20)
    at Module._compile (internal/modules/cjs/loader.js:1138:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1158:10)
loading react.production
Trace: 
    at Object../node_modules/react/cjs/react.production.min.js (/Users/misiek/dev/gatsby-starter-blog/public/webpack:/lib/node_modules/react/cjs/react.production.min.js:29:1)
    at require (/Users/misiek/dev/gatsby-starter-blog/public/webpack:/lib/webpack/bootstrap:19:1)
    at Module../.cache/ssr-develop-static-entry.js (/Users/misiek/dev/gatsby-starter-blog/public/webpack:/lib/.cache/ssr-develop-static-entry.js:14:11)
    at __webpack_require__ (/Users/misiek/dev/gatsby-starter-blog/public/webpack:/lib/webpack/bootstrap:19:1)
    at /Users/misiek/dev/gatsby-starter-blog/public/webpack:/lib/webpack/bootstrap:83:1
    at /Users/misiek/dev/gatsby-starter-blog/public/render-page.js:97:10
    at webpackUniversalModuleDefinition (/Users/misiek/dev/gatsby-starter-blog/public/webpack:/lib/webpack/universalModuleDefinition:3:1)
    at Object.<anonymous> (/Users/misiek/dev/gatsby-starter-blog/public/webpack:/lib/webpack/universalModuleDefinition:10:2)
    at Module._compile (internal/modules/cjs/loader.js:1138:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1158:10)
```

Reply:
haha ok ok you win 😆 

Yeah didn't think this through very well. Thanks for investigating!